### PR TITLE
Bring back MXNET_GPU_COPY_NTHREADS env variable

### PR DESCRIPTION
**Pull request by ctcyang**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Restores MXNET_GPU_COPY_NTHREADS environmental variable
- [x] This variable has been set to MXNET_GPU_WORKER_THREADS (default = 2) since MXNet v0.10.0, so env_var doc is changed to reflect that

## Comments ##
- This MXNET_GPU_COPY_NTHREADS environmental variable was (accidentally) removed in commit 3b517f1 with the value taking the value of MXNET_GPU_WORKER_NTHREADS. Then, shortly afterwards the member variable was itself removed in this PR: https://github.com/apache/incubator-mxnet/pull/6773. Since then, it has been set to MXNET_GPU_WORKER_NTHREADS, which is 2.
- Since this variable controls how many CUDA streams are launched in order to do copying from GPU to GPU, it might be good to bring it back so the user has another knob they can use to tune single-machine multi-GPU training.
- Some experimental evidence that 2 may be better than 1. From running `tools/bandwidth/measure.py` on ResNet-50 with 8 GPUs.

```
export MXNET_GPU_COPY_NTHREADS=1
INFO:root:iter 1, 0.033930 sec, 5.261591 GB/sec per gpu, error 0.000000
INFO:root:iter 2, 0.033478 sec, 5.332636 GB/sec per gpu, error 0.000000
INFO:root:iter 3, 0.044507 sec, 4.011239 GB/sec per gpu, error 0.000000
INFO:root:iter 4, 0.034598 sec, 5.159997 GB/sec per gpu, error 0.000000
INFO:root:iter 5, 0.038231 sec, 4.669686 GB/sec per gpu, error 0.000000
export MXNET_GPU_COPY_NTHREADS=2
INFO:root:iter 1, 0.031341 sec, 5.696318 GB/sec per gpu, error 0.000000
INFO:root:iter 2, 0.031575 sec, 5.654080 GB/sec per gpu, error 0.000000
INFO:root:iter 3, 0.035124 sec, 5.082800 GB/sec per gpu, error 0.000000
INFO:root:iter 4, 0.030341 sec, 5.884048 GB/sec per gpu, error 0.000000
INFO:root:iter 5, 0.032198 sec, 5.544681 GB/sec per gpu, error 0.000000
export MXNET_GPU_COPY_NTHREADS=4
INFO:root:iter 1, 0.046252 sec, 3.859864 GB/sec per gpu, error 0.000000
INFO:root:iter 2, 0.034504 sec, 5.174116 GB/sec per gpu, error 0.000000
INFO:root:iter 3, 0.039854 sec, 4.479529 GB/sec per gpu, error 0.000000
INFO:root:iter 4, 0.034815 sec, 5.127841 GB/sec per gpu, error 0.000000
INFO:root:iter 5, 0.034434 sec, 5.184613 GB/sec per gpu, error 0.000000
```

### CODE REVIEW

File: src/engine/threaded_engine_perdevice.cc [hunk 1]
```diff
@@ -194,6 +196,8 @@ class ThreadedEnginePerDevice : public ThreadedEngine {
   size_t cpu_worker_nthreads_;
   /*! \brief number of concurrent thread each gpu worker uses */
   size_t gpu_worker_nthreads_;
+  /*! \brief number of concurrent thread each gpu copy worker uses */
+  int gpu_copy_nthreads_;
```
Comment (rahul003) [inline src/engine/threaded_engine_perdevice.cc]: Make this size_t as well to maintain uniformity

Comment (ctcyang) [reply src/engine/threaded_engine_perdevice.cc]: Fixed

**Comment by piiswrong** [comment]
Why use 2 by default?

**Comment by ctcyang** [comment]
I set it to use 2 by default, because it has been set to 2 by default since v0.1.0, so I didn't want to cause performance degradation out of the box for users.

Performance-wise, 2 has better KVStore performance than 1 on 8 GPUs (p3.16x instance) for ResNet-50, VGG-16, Inception and AlexNet. 4 is better than 2, but only in some cases (VGG and AlexNet). I measured the bandwidth using `tools/bandwidth/measure.py`. This latency/bandwidth measurement has a direct correlation with end-to-end performance. Higher the bandwidth, higher the throughput end-to-end.

**ResNet:**
```
export MXNET_GPU_COPY_NTHREADS=1
INFO:root:iter 1, 0.033930 sec, 5.261591 GB/sec per gpu, error 0.000000
INFO:root:iter 2, 0.033478 sec, 5.332636 GB/sec per gpu, error 0.000000
INFO:root:iter 3, 0.044507 sec, 4.011239 GB/sec per gpu, error 0.000000
INFO:root:iter 4, 0.034598 sec, 5.159997 GB/sec per gpu, error 0.000000
INFO:root:iter 5, 0.038231 sec, 4.669686 GB/sec per gpu, error 0.000000
export MXNET_GPU_COPY_NTHREADS=2
INFO:root:iter 1, 0.031341 sec, 5.696318 GB/sec per gpu, error 0.000000
INFO:root:iter 2, 0.031575 sec, 5.654080 GB/sec per gpu, error 0.000000
INFO:root:iter 3, 0.035124 sec, 5.082800 GB/sec per gpu, error 0.000000
INFO:root:iter 4, 0.030341 sec, 5.884048 GB/sec per gpu, error 0.000000
INFO:root:iter 5, 0.032198 sec, 5.544681 GB/sec per gpu, error 0.000000
export MXNET_GPU_COPY_NTHREADS=4
INFO:root:iter 1, 0.046252 sec, 3.859864 GB/sec per gpu, error 0.000000
INFO:root:iter 2, 0.034504 sec, 5.174116 GB/sec per gpu, error 0.000000
INFO:root:iter 3, 0.039854 sec, 4.479529 GB/sec per gpu, error 0.000000
INFO:root:iter 4, 0.034815 sec, 5.127841 GB/sec per gpu, error 0.000000
INFO:root:iter 5, 0.034434 sec, 5.184613 GB/sec per gpu, error 0.000000
```

**VGG:**
```
export MXNET_GPU_COPY_NTHREADS=1
INFO:root:iter 1, 0.334008 sec, 2.899644 GB/sec per gpu, error 0.000000
INFO:root:iter 2, 0.328484 sec, 2.948406 GB/sec per gpu, error 0.000000
INFO:root:iter 3, 0.335069 sec, 2.890458 GB/sec per gpu, error 0.000000
INFO:root:iter 4, 0.330471 sec, 2.930672 GB/sec per gpu, error 0.000000
INFO:root:iter 5, 0.332817 sec, 2.910013 GB/sec per gpu, error 0.000000
export MXNET_GPU_COPY_NTHREADS=2
INFO:root:iter 1, 0.292127 sec, 3.315352 GB/sec per gpu, error 0.000000
INFO:root:iter 2, 0.296561 sec, 3.265785 GB/sec per gpu, error 0.000000
INFO:root:iter 3, 0.287905 sec, 3.363961 GB/sec per gpu, error 0.000000
INFO:root:iter 4, 0.292065 sec, 3.316053 GB/sec per gpu, error 0.000000
INFO:root:iter 5, 0.288955 sec, 3.351746 GB/sec per gpu, error 0.000000
export MXNET_GPU_COPY_NTHREADS=4
INFO:root:iter 1, 0.269719 sec, 3.590786 GB/sec per gpu, error 0.000000
INFO:root:iter 2, 0.272700 sec, 3.551528 GB/sec per gpu, error 0.000000
INFO:root:iter 3, 0.274158 sec, 3.532648 GB/sec per gpu, error 0.000000
INFO:root:iter 4, 0.280010 sec, 3.458815 GB/sec per gpu, error 0.000000
INFO:root:iter 5, 0.268548 sec, 3.606448 GB/sec per gpu, error 0.000000
```

**Inception:**
```
export MXNET_GPU_COPY_NTHREADS=1
INFO:root:iter 1, 0.039140 sec, 4.256541 GB/sec per gpu, error 0.000000
INFO:root:iter 2, 0.044692 sec, 3.727735 GB/sec per gpu, error 0.000000
INFO:root:iter 3, 0.039852 sec, 4.180527 GB/sec per gpu, error 0.000000
INFO:root:iter 4, 0.043019 sec, 3.872746 GB/sec per gpu, error 0.000000
INFO:root:iter 5, 0.041909 sec, 3.975259 GB/sec per gpu, error 0.000000
export MXNET_GPU_COPY_NTHREADS=2
INFO:root:iter 1, 0.041007 sec, 4.062764 GB/sec per gpu, error 0.000000
INFO:root:iter 2, 0.040754 sec, 4.087982 GB/sec per gpu, error 0.000000
INFO:root:iter 3, 0.040930 sec, 4.070360 GB/sec per gpu, error 0.000000
INFO:root:iter 4, 0.037946 sec, 4.390421 GB/sec per gpu, error 0.000000
INFO:root:iter 5, 0.040932 sec, 4.070147 GB/sec per gpu, error 0.000000
export MXNET_GPU_COPY_NTHREADS=4
INFO:root:iter 1, 0.054199 sec, 3.073849 GB/sec per gpu, error 0.000000
INFO:root:iter 2, 0.049373 sec, 3.374356 GB/sec per gpu, error 0.000000
INFO:root:iter 3, 0.047462 sec, 3.510233 GB/sec per gpu, error 0.000000
INFO:root:iter 4, 0.046725 sec, 3.565597 GB/sec per gpu, error 0.000000
INFO:root:iter 5, 0.048453 sec, 3.438414 GB/sec per gpu, error 0.000000
```

**AlexNet**
```
export MXNET_GPU_COPY_NTHREADS=1
INFO:root:iter 1, 0.108810 sec, 3.270900 GB/sec per gpu, error 0.000000
INFO:root:iter 2, 0.107738 sec, 3.303473 GB/sec per gpu, error 0.000000
INFO:root:iter 3, 0.111432 sec, 3.193953 GB/sec per gpu, error 0.000000
INFO:root:iter 4, 0.110318 sec, 3.226196 GB/sec per gpu, error 0.000000
INFO:root:iter 5, 0.113937 sec, 3.123716 GB/sec per gpu, error 0.000000
export MXNET_GPU_COPY_NTHREADS=2
INFO:root:iter 1, 0.101540 sec, 3.505091 GB/sec per gpu, error 0.000000
INFO:root:iter 2, 0.096378 sec, 3.692832 GB/sec per gpu, error 0.000000
INFO:root:iter 3, 0.094568 sec, 3.763495 GB/sec per gpu, error 0.000000
INFO:root:iter 4, 0.099746 sec, 3.568153 GB/sec per gpu, error 0.000000
INFO:root:iter 5, 0.098334 sec, 3.619385 GB/sec per gpu, error 0.000000
export MXNET_GPU_COPY_NTHREADS=4
INFO:root:iter 1, 0.089701 sec, 3.967718 GB/sec per gpu, error 0.000000
INFO:root:iter 2, 0.091634 sec, 3.884036 GB/sec per gpu, error 0.000000
INFO:root:iter 3, 0.087382 sec, 4.073010 GB/sec per gpu, error 0.000000
INFO:root:iter 4, 0.090852 sec, 3.917448 GB/sec per gpu, error 0.000000
INFO:root:iter 5, 0.090665 sec, 3.925535 GB/sec per gpu, error 0.000000
```

**Comment by marcoabreu** [comment]
Would it make sense to choose this value dependent on the number of GPUs available?

**Comment by ctcyang** [comment]
When it's set to 1, copies between GPUs are serialized. Even between 2 GPUs, if GPU 0 is currently sending something to 1 on one stream, and GPU 1 wants to send something to 0, the latter send will have to wait leaving the 1 -> 0 link under-utilized.

**Pull request merged**